Aikido.Zen.Core/Models/Ip/Blocklist.cs

```diff
@@ -137,17 +181,92 @@ public bool IsIPAllowed(string ip, string endpoint)
+                    return true; // Invalid endpoint format, allow by default
                 }
 
-                if (!trie.HasItems)
+                var method = parts[0];
+                var path = parts[1];
+                var context = new Context { Method = method, Route = path };
+
+                var matchingEndpoints = RouteHelper.MatchEndpoints(context, _endpointConfigs.Select(x => x.Config).ToList());
+                if (!matchingEndpoints.Any())
                 {
                     return true; // Allow if no specific subnets are defined for the endpoint
                 }
 
-                return trie.IsIpInRange(ip);
+                // Check if any matching endpoint allows the IP
+                foreach (var matchingEndpoint in matchingEndpoints.Where(e => (e.AllowedIPAddresses?.Count() ?? 0) > 0))
+                {
+                    var endpointConfig = _endpointConfigs.First(x => x.Config == matchingEndpoint);
+                    if (endpointConfig.AllowedIPs.HasItems)
+                    {
+                        if (!endpointConfig.AllowedIPs.IsIpInRange(ip))
+                        {
+                            return false; // If IP is not allowed for this endpoint, return false
+                        }
+                    }
+                }
+
+                return true; // If no endpoints disallow the IP, return true
+            }
+            finally
+            {
+                _lock.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        /// Checks if an IP is bypassed.
+        /// </summary>
+        /// <param name="ip">The IP address to check.</param>
+        /// <returns>True if the IP is bypassed, false otherwise.</returns>
+        public bool IsIPBypassed(string ip)
+        {
+            _lock.EnterReadLock();
+            try
+            {
+                if (!IPHelper.IsValidIp(ip))
+                {
+                    return false; // Invalid IPs are not allowed, since allowing bypasses other blocking rules
+                }
+
+                if (_bypassedIps.HasItems)
+                {
+                    return _bypassedIps.IsIpInRange(ip);
+                }
+
+                return false;
+            }
+            finally
+            {
+                _lock.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        /// Checks if an IP is allowed.
+        /// </summary>
+        /// <param name="ip">The IP address to check.</param>
+        /// <returns>True if the IP is allowed, false otherwise.</returns>
+        public bool IsIPAllowed(string ip)
+        {
+            _lock.EnterReadLock();
+            try
+            {
+                if (!IPHelper.IsValidIp(ip))
+                {
+                    return !_allowedIps.HasItems; // Invalid IPs are not allowed if there are allowed IPs
+                }
+
+                if (_allowedIps.HasItems)
```

Comment:
Let's put this check first, makes it cleaner and easier to read (also means we don't have to validate IPs when we don't need to)